### PR TITLE
blog: using mongodb with node-red

### PR DESCRIPTION
**Pull request by sumitshinde-84**
## Checklist

 - [ ] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] I have considered the performance impact of these changes
 - [ ] Suitable unit/system level tests have been added and they pass
 - [ ] Documentation has been updated

### CODE REVIEW

**Comment by sumitshinde-84** [comment]
PR  is not yet ready for review

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+<!--more-->
+
+## What is MongoDB
+
+MongoDB is an open-source, NoSQL database system designed for flexibility and scalability in handling data for modern applications. It stores data in a format called JSON-like documents, which offers a schema-less approach where each document can have its unique structure. This flexibility allows for easy schema evolution and dynamic querying. MongoDB is widely used in web and mobile applications for its ability to handle large volumes of data and support real-time applications efficiently.
+
+In MongoDB, instead of tables, rows, and columns as in SQL databases, we have collections, documents, and fields.
+
+!["Annotomy of MongoDB document"](./images/using-mongodb-with-node-red-annotomy-of-mongodb-document.png "Annotomy of MongoDB document"){data-zoomable}
```
Comment (sumitshinde-84) [inline src/blog/2024/04/using-mongodb-with-node-red.md:25]: @ZJvandeWeg Should we recreate this graphic with our ff colors? I downloaded it from the internet.

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
+---
+title: Using Mongodb With Node-RED
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:2]:
```suggestion
title: Using MongoDB With Node-RED
```

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+subtitle: Guide to integrating MongoDB database into Node-RED Applications, for retrieving, updating, and deleting data
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:3]:
```suggestion
subtitle: Guide to integrating MongoDB database into Node-RED Applications, for writing, updating, and deleting data.
```

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+description: Learn how to seamlessly integrate MongoDB, a NoSQL database, into your Node-RED applications with this step-by-step guide. Gain a fundamental understanding of MongoDB's key concepts and terminologies. 
+date: 2024-04-05
+authors: ["sumit-shinde"]
+image:
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:7]: Art request please :)

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
+---
+title: Using Mongodb With Node-RED
+subtitle: Guide to integrating MongoDB database into Node-RED Applications, for retrieving, updating, and deleting data
+description: Learn how to seamlessly integrate MongoDB, a NoSQL database, into your Node-RED applications with this step-by-step guide. Gain a fundamental understanding of MongoDB's key concepts and terminologies. 
+date: 2024-04-05
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:5]:
```suggestion
date: 2024-04-12
```

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+    - NoSQL
+---
+
+MongoDB, a popular NoSQL database, is a favored choice among developers. As MongoDB continuously releases updates to enhance efficiency and suitability for manufacturing purposes, we find it essential to provide content for our Node-RED users. In this guide, I will cover everything from MongoDB foundations to setup, providing you with step-by-step instructions for integration into your Node-RED applications.
+
+<!--more-->
+
+## What is MongoDB
+
+MongoDB is an open-source, NoSQL database system designed for flexibility and scalability in handling data for modern applications. It stores data in a format called JSON-like documents, which offers a schema-less approach where each document can have its unique structure. This flexibility allows for easy schema evolution and dynamic querying. MongoDB is widely used in web and mobile applications for its ability to handle large volumes of data and support real-time applications efficiently.
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:21]:
```suggestion
MongoDB is an open-source, NoSQL database system designed for flexibility and scalability in handling data for modern applications. It stores data in documents, which are JSON-like structures. MongoDB offers a schema-less approach where each document can have its unique structure. This flexibility allows for easy schema evolution and dynamic querying. MongoDB is widely used in web and mobile applications for its ability to handle large volumes of data and support real-time applications efficiently.
```

File: src/blog/2024/04/using-mongodb-with-node-red.md (hunk 1 above)
Comment (ZJvandeWeg) [reply src/blog/2024/04/using-mongodb-with-node-red.md:25]: Please add it to the art request I mentioned earlier

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+### Installing MongoDB custom node
+
+- Click the Node-RED Settings (top-right)
+- Click "Manage Palette"
+- Switch to the "Install" tab
+- Search for `node-red-contrib-mongodb4`
+
+### Understanding MongoDB Configuration
+
+To establish a connection between our Node-RED application and the MongoDB database, it's essential to have the following configuration details prepared ( While writing the article, I have utilized MongoDB Atlas, but feel free to choose any MongoDB setup, as the instructions in the article are adaptable for any setup ) :
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:47]:
```suggestion
While writing the article, I have utilized MongoDB Atlas, but feel free to choose any MongoDB setup, as the instructions in the article are adaptable for any setup. To establish a connection between our Node-RED application and the MongoDB database, it's essential to have the following configuration details prepared:
```

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+- Host: IP address or hostname of your MongoDB server.
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:49]:
```suggestion
- `Host`: IP address or hostname of your MongoDB server.
```

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+- Port: By default, MongoDB uses port 27017. ( If you're using a managed MongoDB service like MongoDB Atlas, this information may not be required.)
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:50]:
```suggestion
- `Port`: By default, MongoDB uses port 27017. ( If you're using a managed MongoDB service like MongoDB Atlas, this information may not be required.)
```

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+- Database: The name of the MongoDB database you want to connect to.
+- User: Username with the necessary privileges to access the specified database.
+- Password: Corresponding password for the username.
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:53]:
```suggestion
- `Database`: The name of the MongoDB database you want to connect to.
- `User`: Username with the necessary privileges to access the specified database.
- `Password`: Corresponding password for the username.
```

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+
+!["Screenshot displaying connection configuration of MongoDB 4 node."](./images/using-mongodb-with-node-red-mongodb-node-connection-configuration.png "Screenshot displaying connection configuration of MongoDB 4 node."){data-zoomable}
+
+## Using MongoDB with Node-RED
+
+In this section, we'll delve into leveraging MongoDB to perform operations such as storing, retrieving, updating, and deleting data within Node-RED. To enhance understanding and engagement, we'll develop a basic customer relationship manager system utilizing MongoDB and Node-RED. 
+
+For the CRM system we will develop, the data structure will be as follows:
+
+```json
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:83]: I think you need some `{% raw %}` tags?

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+- Drag the ui-form widget onto the canvas, create a new ui-group for it, and add form elements for "id" and "status".
+
+!["Screenshot displaying form widget configuration to update data in MongoDB"](./images/using-mongodb-with-node-red-insert-data-form.png "Screenshot displaying form widget configuration to update data in MongoDB"]{data-zoomable}
+
+- Drag the change node onto the canvas. In the change node, set `msg.payload` as the following JSON object:
+
+```json
+[
+    // Specifying the criteria to identify the document to update, based on its _id field
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:171]: Not sure if we should allow people to copy+paste illegal JSON?

**Review by ZJvandeWeg — CHANGES_REQUESTED** [review]

File: src/blog/2024/04/using-mongodb-with-node-red.md (hunk 12 above)
Comment (sumitshinde-84) [reply src/blog/2024/04/using-mongodb-with-node-red.md:171]: How about if we add the following note before code blocks:
Note:- *Please note that the comments provided are for explanation purposes only and should not be copied along with the code*

File: src/blog/2024/04/images/using-mongodb-with-node-red-mongodb-deleteone-flow.png
Comment (ZJvandeWeg) [inline src/blog/2024/04/images/using-mongodb-with-node-red-mongodb-deleteone-flow.png:1]: Let's deploy the flow next time before the screenshot so it removes the blue dot

File: src/blog/2024/04/using-mongodb-with-node-red.md (hunk 4 above)
Comment (ZJvandeWeg) [reply src/blog/2024/04/using-mongodb-with-node-red.md:7]: @sumitshinde-84 Can you link the issue in this thread?

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+
+This section contains helpful links to other content. It isn’t required, so consider it supplemental.
+
+- [MongoDB best practices](https://www.mongodb.com/basics/best-practices) for to increase performance
+- [Ebook: MongoDB in manufacturing](https://www.mongodb.com/collateral/manufacturing-with-mongodb).
+- [Manufacturing & Automotive Industry Knowledge Accelerator](https://learn.mongodb.com/learning-paths/manufacturing-automotive-industry-knowledge-accelerator)
+
+## Conclusion
+
+In this guide, we've covered the integration of MongoDB with Node-RED to create dynamic applications. We've walked through the foundation, setup process, demonstrated how to perform CRUD operations, and discussed debugging techniques. Additionally, to enhance understanding and engagement we've built a Customer Relationship Management system.
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:281]: Should we add a link to the blueprint that's available too?

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+
+- Connect the nodes' wires as shown in the below image.
+
+!["Screenshot displaying connections of wires in the 'Drop collecton from Database' flow"](./images/using-mongodb-with-node-red-mongodb-drop-flow.png "Screenshot displaying connections of wires in the 'Drop collecton from Database' flow"){data-zoomable}
+
+## Debugging the operations
+
+To debug MongoDB operations in Node-RED, simply add a debug node after the MongoDB4 node. This enables you to monitor the execution of operations and diagnose any errors that may occur. Below are sample debug messages indicating successful operations for each operation we have covered in this guide.
+
+```
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:235]:
```suggestion
```js
```

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+
+### Deleting Customer Data to Database
+
+- Drag the ui-form widget onto the canvas, create a new ui-group for it, and add form elements for id and name.
+
+!["Screenshot displaying form widget configuration to delete data in MongoDB"](./images/using-mongodb-with-node-red-delete-data-form.png "Screenshot displaying form widget configuration to delete data in MongoDB"){data-zoomable}
+
+- Drag the change node onto the canvas. In the change node, set `msg.payload` as the following JSON object:
+
+```json
```
Comment (ZJvandeWeg) [inline src/blog/2024/04/using-mongodb-with-node-red.md:196]: As YAML is a superset of JSON, we can fix the highlighting with:

```suggestion
```yaml
```

File: src/blog/2024/04/using-mongodb-with-node-red.md (hunk 12 above)
Comment (ZJvandeWeg) [reply src/blog/2024/04/using-mongodb-with-node-red.md:171]: Not as a comment in the code I think, just above it?

**Review by ZJvandeWeg — CHANGES_REQUESTED** [review]

File: src/blog/2024/04/using-mongodb-with-node-red.md (hunk 13 above)
Comment (sumitshinde-84) [reply src/blog/2024/04/using-mongodb-with-node-red.md:281]: Yes, let's include the link to the blueprint.

File: src/blog/2024/04/using-mongodb-with-node-red.md [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+!["Screenshot displaying flow of CRM System"](./images/using-mongodb-with-node-red-crm-system-node-red-flow.png "Screenshot displaying flow of CRM System"){data-zoomable}
+
+- With your flow updated to include the above, click the "Deploy" button in the top-right of the Node-RED Editor.
+- Locate the 'Open Dashboard' button at the top-right corner of the Dashboard 2.0 sidebar and click on it to navigate to the dashboard.
+
+Now To drop tables, and retrieve table data, click on the 'Inject Node' button. For customer data insertion, updates, and deletions, navigate to fill in the corresponding forms and submit.
+
+!["Screenshot displaying dashboard view of CRM System"](./images/using-mongodb-with-node-red-crm-system-node-red-dashboard-view.png "Screenshot displaying dashboard view of CRM System"){data-zoomable}
+
+## Additional resources
```
Comment (sumitshinde-84) [inline src/blog/2024/04/using-mongodb-with-node-red.md:273]: What do you think about this new section? Should we add it to our next upcoming post as well?

File: src/blog/2024/04/images/using-Mongo-with-nr.png
Comment (ZJvandeWeg) [inline src/blog/2024/04/images/using-Mongo-with-nr.png:1]: <3

Comment (sumitshinde-84) [reply src/blog/2024/04/images/using-Mongo-with-nr.png:1]: Can we use this graphic for the blog tile and then a cropped version of the same graphic to explain the anatomy of MongoDB documents, or should we use two different images?

**Review by ZJvandeWeg — APPROVED** [review]

**Pull request merged**